### ISSUE TITLE
RSA padding oracle issue

### ISSUE THREAD

**Issue by ionut-arm** [parallaxsecond/parsec#619]
I've discovered a bug that can turn out to be a security vulnerability for Parsec clients that use Parsec in a very specific way.

TL;DR: **If any Parsec client uses the service for TLS 1.2 pre-master secret decryption with RSA PKCS#1 v1.5, they're vulnerable to Bleichenbacher's padding oracle attack.**

## The bug

The Parsec service, following the lead of the PSA Crypto, supports two algorithms for asymmetric encryption/decryption: RSA OAEP, and RSA PKCS#1 v1.5. While both algorithms deal with padding, the current issue only applies to RSA PKCS#1 v1.5, so this issue touches on the use of that one only.

Whenever we complete an `psa_asymmetric_decrypt` operation, we perform a conversion from the provider-specific error code to the Parsec-specific error code (i.e., the code that will be sent to the Parsec client). Given that these error codes are not exactly "standardised" across the providers, we decided (on first implementation) to properly map only the codes that seem most relevant and that have an equivalent in the PSA Crypto API, and for the rest to return a generic error. The bug concerns these mappings (or lack thereof) - we do not properly map the error codes signifying an invalid padding to `PsaErrorInvalidPadding` for the TPM and PKCS11 providers. This means that Parsec clients would instead get a generic error code instead when a padding is found to be invalid.

## The security issue

> Ok, so why is an invalid padding important in the case of RSA PKCS#1 v1.5 encryption?

RSA PKCS#1 v1.5 encryption is vulnerable to a type of attack that was discovered by Bleichenbacher back in '98 (see a summary [here](https://cryptosense.com/blog/why-pkcs1v1-5-encryption-should-be-put-out-of-our-misery)). The attacker leverages knowledge of when a ciphertext they sent produced an invalid padding when decrypted (i.e., a padding oracle) to break the system and recover the private RSA key of the peer.

> Well then it's good that we don't reveal information about when the padding is incorrect, right?

The padding oracle problem is relevant in the context of TLS - TLS 1.2 allows the client to unilaterally produce a Pre-Master Secret (PMS), encrypt it with the server's public key, and continue to use that PMS to secure the channel post-handshake. Thus, if PKCS#1 v1.5 is used, the server must take some precautions to never reveal that the PMS decryption resulted in an invalid pad. Instead, the server [should just choose a random value R as a placeholder and pretend that nothing bad happened](https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.7.1). (This will result in an error later on in the handshake when the PMS is actually supposed to be used.) The problem with our bug is: the server cannot perform this mitigation if it doesn't know the error was caused by an invalid padding. Generic errors in our backends are most likely very infrequent, so the confusion between "padding error" and "generic error" can easily be used as a padding oracle.

## The fix

All we need to do is to make sure we always report back when a padding error has just happened. We should also test for this.

**Comment by ionut-arm** [comment]
This issue puts Parsec in a tricky situation, particularly with regards to new providers in the future.

RSA decryption providers and implementations (such as the Parsec service) have to report back some error code when a decryption fails. The error could either be "benign" (something from which an attacker cannot learn anything about the key or plaintext), or "malign" (from which they could learn). If the returned error code is always the same, generic over all errors, then the user/client has no choice but to assume the worst and perform whatever mitigation are expected for the "malign" case. If the returned error code can distinguish between the two types of errors, then clients can go ahead and perform different steps based on that (for example, implementing the mitigations only when a "malign" error occurs). Parsec is in the latter category - we send different response codes for different problems.

Now, the problem is that Parsec may not always be able to properly satisfy that contract. If a backend does not differentiate between the error types, then we (1) send out a generic error, or (2) send out a "malign" error code. For (1), the client could end up in a pickle, not performing the required mitigations. For (2), our reported error code will technically be incorrect, probably a lot of the time. (1) is obviously worse, but (2) isn't that good either.

And then there are things like PKCS11, for which different implementations can do different things. SoftHSMv2, for example, always replies with `GENERIC_ERROR` when decryption fails (option (1) from above), even though a specific error code for "malign" errors exists.

It's a mess.
